JavaScript/chapter01/p02_check_perm/avc278.js

```diff
@@ -0,0 +1,98 @@
+
+/**
+ * Checks if a string is a permutation of another
+ * @param  {string} A source input string to check against
+ * @param  {string} B target input string to compare with source string
+ * @return {bool}     Whether the input strings are permutations of one another
+ *
+ * Throughout the problem, we set up a couple ways to early exit. At the start, if both strings are not of the same
+ * length, we already know they cannot be permutations of each other, so we return false.
+ * Otherwise, we can continue and set up a mapping between the elements in A and B, and their frequencies.
+ * This is an O(N) operation with respect to time and space where N is the length of the string A.
+ * At this point, we can get the keys from the frequency objects we created, which is an O(N) operation for space.
+ * If the lengths of `aKeys` and `bKeys` are not the same, we know they cannot be permutations and we exit early.
+ * Otherwise, we can continue to iterate through the keys. We need to make sure that the frequencies of aKeys exist in
+ * both aFreq and bFreq. If at any point, they do not match, we can return false.
+ *
+ * If we haven't returned false up to this point, it's safe to return true.
+ *
+ * Runtime: O(N)
+ * Space:   O(N)
+ *
+ */
+const checkPermutation1 = (A, B) => {
+  if (A.length !== B.length) return false;
+  const aFreq = {};
+  const bFreq = {};
+  for (const a of A) {
+    aFreq[a] = (aFreq[a] || 0) + 1;
+  }
+  for (const b of B) {
+    bFreq[b] = (bFreq[b] || 0) + 1;
+  }
+
+  const aKeys = Object.keys(aFreq);
+  const bKeys = Object.keys(bFreq);
+  if (aKeys.length !== bKeys.length) return false;
+  for (let i = 0; i < aKeys.length; i++) {
+    if (aFreq[aKeys[i]] !== bFreq[aKeys[i]]) return false;
+  }
+
+  return true;
+};
+
+/**
+ * Checks if a string is a permutation of another
+ * @param  {string} A source input string to check against
+ * @param  {string} B target input string to compare with source string
+ * @return {bool}     Whether the input strings are permutations of one another
+ *
+ * An alternative to this problem, and significantly more elegant, is to simply split the input strings into arrays,
+ * use the built-in sort array method, then join back to a string, and check if A and B are equal.
+ * The trade off is since we sort arrays of length N, the runtime becomes slower from O(N) to O(NlogN).
+ * As for space, we create arrays of length N, so there is no change to space performance compared to checkPermutation1.
+ * Even though this is less performant to checkPermutation1, for production code, this seems to be more maintainable
+ * and way easier to read at a glance.
+ *
+ * Runtime: O(NlogN)
+ * Space:   O(N)
+ *
+ */
+const checkPermutation2 = (A, B) => {
+  if (A.length !== B.length) return false;
+
+  A = A.split("").sort().join("");
+  B = B.split("").sort().join("");
+
+  return A === B;
+};
+
+describe(module.filename, () => {
+  it("should return false on input strings not of the same size", () => {
```

Comment:
minor optional note: it could be better to use [parameterized tests](https://mochajs.org/#dynamically-generating-tests) to avoid repetition:

```javascript
  const checkPermutations = [checkPermutation1, checkPermutation2];
  checkPermutations.forEach((checkPermutation) => {
    describe(checkPermutation.name, () => {
      it("should return false on input strings not of the same size", () => {
        assert.ok(!checkPermutation("hello", "hi"));
      });
      // ...
    });
  });
```

Reply:
Oh interesting. I'll take a look at it in a bit and update my tests